packages/gestalt/src/SideNavigation/navigationChildrenUtils.js

```diff
@@ -0,0 +1,113 @@
+// @flow strict
+import { type Node as ReactNode } from 'react';
+import classnames from 'classnames';
+import ItemsEllipsis, { type Props as EllipsisProps } from './ItemsEllipsis';
+import styles from '../SideNavigation.css';
+import flattenChildren, { type ReactChildArray } from '../utils/flattenChildren';
+
+export const ALLOWED_CHILDREN_MAP = {
+  main: ['SideNavigation.Section', 'SideNavigation.TopItem', 'SideNavigation.Group'],
+  nested: ['SideNavigation.NestedItem', 'SideNavigation.NestedGroup'],
+};
+
+export function validateChildren({
+  children,
+  filterLevel,
+}: {
+  children: ReactChildArray,
+  filterLevel: 'main' | 'nested',
+}) {
+  children.forEach((child) => {
+    const isTopLevel = filterLevel === 'main';
+
+    if (!ALLOWED_CHILDREN_MAP[filterLevel].includes(child.type.displayName)) {
+      throw new Error(
+        `Gestalt ${child.type.displayName} cannot be used at ${
+          isTopLevel ? 'the top' : 'a nested'
+        } level`,
+      );
+    }
+  });
+}
+
+export function countItemsWithIcon(children: ReactChildArray): number {
+  return flattenChildren(children).reduce(
+    (count, child) =>
+      child.type.displayName === 'SideNavigation.Section'
+        ? count + countItemsWithIcon(child.props.children)
+        : count + Number(Boolean(child.props.icon)),
+    0,
+  );
+}
+
+export function getChildrenActiveProp(
+  children: ReactChildArray,
+): $PropertyType<EllipsisProps, 'active'> {
+  if (children.length === 0) return undefined;
+
+  const activeChild = children.find((child) => !!child.props?.active);
+
+  if (activeChild?.props?.active) return activeChild.props.active;
+
+  const grandChildren = children
+    .filter(
+      (child) =>
+        child?.type?.displayName === 'SideNavigation.Group' ||
+        child?.type?.displayName === 'SideNavigation.NestedGroup',
+    )
+    .map((child) => flattenChildren(child?.props?.children))
+    .flat();
+
+  return getChildrenActiveProp(grandChildren);
+}
+
+function renderEllipses(items: $ReadOnlyArray<React$Element<empty> | EllipsisProps>) {
+  return items.map((item, i) => {
+    if (item.props) return item;
+
+    return (
+      // eslint-disable-next-line react/no-array-index-key
+      <li key={i} className={classnames(styles.liItem)}>
+        <ItemsEllipsis {...item} />
+      </li>
+    );
+  });
+}
+
+export function groupIconlessChildren(children: ReactChildArray): $ReadOnlyArray<ReactNode> {
```

Comment:
This function is a bit crazy one. It is for the items that are not grouped in a `Section`. 

Creates ellipses for each portion of "non-section" items (separated by `Section`s) if there are items without icons. 
That means, If there are "non-section" items before and after `Section`, each will have an ellipsis for iconless items.

Take a look at the [collapsibleWithMixedIconsExample.js](https://github.com/pinterest/gestalt/pull/3476/files#diff-f0f42099ab0c6407c1cfe8fb580d5f27ebb11d5a4004e28493796639be948fc7) file for the usecase.

Reply:
Also, if there are items with 

1. notification labels
2. `active` prop

algorithms takes the 1st one's infor and sets it to the ellipsis.

Reply:
can we add comments in the code so it's easier to read the code?